### ISSUE TITLE
Issue of Object number in Inference

### ISSUE THREAD

**Issue by QinghongLin** [hkchengrex/Cutie#31]
Hi, Thanks for the impressive work and codes!

I found there might has an issue during inference stage. https://github.com/hkchengrex/Cutie/blob/main/cutie/inference/inference_core.py#L281

I wondering the `if tmp_id >= pred_prob_no_bg.shape[0]:` should be `if tmp_id + 1 >= pred_prob_no_bg.shape[0]:`?
```
                    if tmp_id >= pred_prob_no_bg.shape[0]:
                        new_masks.append(this_mask.unsqueeze(0))
                    else:
                        # +1 for padding the background channel
                        pred_prob_no_bg[tmp_id + 1] = this_mask
```

When `tmp_id==pred_prob_no_bg-1`; the `pred_prob_no_bg[tmp_id + 1] = this_mask` will raise an error since `tmp_id + 1 (e.g., tmp_id=4, tmp_id+1=5)` already exceeds the channel dimension of `pred_prob_no_bg (5; index start by 0, so the largest accepted index is 4)`

**Comment by QinghongLin** [comment]
after debugging, I think a correct version should be 

```
                new_masks = []
                for mask_id, tmp_id in enumerate(corresponding_tmp_ids):
                    if idx_mask:
                        this_mask = (mask == objects[mask_id]).type_as(pred_prob_no_bg)
                    else:
                        this_mask = mask[tmp_id]
                    # if tmp_id >= pred_prob_no_bg.shape[0]:
                    if tmp_id > pred_prob_no_bg.shape[0]:                        
                        new_masks.append(this_mask.unsqueeze(0))
                    else:
                        # +1 for padding the background channel
                        # pred_prob_no_bg[tmp_id + 1] = this_mask
                        pred_prob_no_bg[tmp_id-1] = this_mask
                # new_masks are always in the order of tmp_id
                mask = torch.cat([pred_prob_no_bg, *new_masks], dim=0)
```

since pred_prob_no_bg tensor do not include the bg channel.

**Comment by hkchengrex** [comment]
Thank you for posting the details. Frankly, this code path is undertested (most evaluations are done with `idx_mask=True`), but it does seem wrong. Have you tested this? If so I can just use your fix. 

Thanks again.

**Comment by QinghongLin** [comment]
@hkchengrex  Hi, yeap, I have tested. I suggest you use this implementation. I can successfully run.
I need to make this modification because I will gradually increase the mask and object class numbers.

```
                new_masks = []
                for mask_id, tmp_id in enumerate(corresponding_tmp_ids):
                    if idx_mask:
                        this_mask = (mask == objects[mask_id]).type_as(pred_prob_no_bg)
                    else:
                        this_mask = mask[tmp_id]
                    # if tmp_id >= pred_prob_no_bg.shape[0]:
                    if tmp_id > pred_prob_no_bg.shape[0]:                        
                        new_masks.append(this_mask.unsqueeze(0))
                    else:
                        # +1 for padding the background channel
                        # pred_prob_no_bg[tmp_id + 1] = this_mask
                        pred_prob_no_bg[tmp_id-1] = this_mask
                # new_masks are always in the order of tmp_id
                mask = torch.cat([pred_prob_no_bg, *new_masks], dim=0)
```

**Comment by hkchengrex** [comment]
Thank you :) 
Merged into main.
